src/Symfony/Component/Utf8/Bytes.php

```diff
@@ -0,0 +1,505 @@
+        setlocale(LC_CTYPE, 'C');
+
+        $result = clone $this;
+        $result->string = $allWords ? ucwords($this->string) : ucfirst($this->string);
+
+        setlocale(LC_CTYPE, $current);
+
+        return $result;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function toFoldedCase(bool $full = true)
+    {
+        return $this->toLowerCase();
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function substr(int $start = 0, int $length = null): self
+    {
+        $result = clone $this;
+        $result->string = substr($this->string, $start, $length);
+
+        return $result;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function trim(string $charsList = null): self
+    {
+        $result = clone $this;
+        $result->string = trim($this->string, $charsList ?: " \t\n\r\0\x0B");
+
+        return $result;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function trimLeft(string $charsList = null): self
+    {
+        $result = clone $this;
+        $result->string = ltrim($this->string, $charsList ?: " \t\n\r\0\x0B");
+
+        return $result;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function trimRight(string $charsList = null): self
+    {
+        $result = clone $this;
+        $result->string = rtrim($this->string, $charsList ?: " \t\n\r\0\x0B");
+
+        return $result;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getIterator(int $maxChunkLength = 1): \Generator
+    {
+        if (1 > $maxChunkLength) {
+            throw new InvalidArgumentException('The maximum length of each segment must be greater than zero.');
+        }
+
+        if ('' === $this->string) {
+            return;
+        }
+
+        foreach (str_split($this->string, $maxChunkLength) as $char) {
+            $clone = clone $this;
+            $clone->string = $char;
+            yield $clone;
+        }
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function append(string $suffix): self
+    {
+        $result = clone $this;
+        $result->string .= $suffix;
+
+        return $result;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function prepend(string $prefix): self
+    {
+        $result = clone $this;
+        $result->string = $prefix.$result->string;
+
+        return $result;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function indexOf(string $needle, int $offset = 0)
+    {
+        if ('' === $needle) {
+            return;
+        }
+
+        if (0 <= $offset || PHP_VERSION_ID >= 70100) {
+            $result = strpos($this->string, $needle, $offset);
+
+            return false === $result ? null : $result;
+        }
+
+        // Workaround to support negative offsets with strpos() in PHP < 7.1
+        $start = $offset + strlen($this->string);
+        $offset = strpos(substr($this->string, $start), $needle);
+
+        return false === $offset ? null : $start + $offset;
+    }
```

Comment:
I read that returning `-1` in JavaScript was "abusing of the type system".
But your point about the "safety" it provides is valid.
On the other side, people are used to check the return value of eg strpos in php, where it returns false|0.
What do others think we should do here?